### ISSUE TITLE
No error reported when using a void value in a list

### ISSUE THREAD

**Issue by nshahan** [dart-lang/language#809]
Both the analyzer and CFE report an error when you try to use a void value in a position like on the line that is commented out in the code below. If you use the value in a list there is no error and in our backends the list contains the implicitly returned null value from the `foo()` method.

```dart
void main() {
  var x = foo();
  // print(x); 
  // Error: This expression has type 'void' and can't be used.

  print([x]); // why ok?
  // Prints: [null]

}

void foo() {}
```

Is this a bug? @leafpetersen @stereotype441 @johnniwinther

**Comment by eernstg** [comment]
`print([x])` where `x` has static type `void` is allowed because `[x]` gets the static type `List<void>`, and there are no constraints on using the result of evaluating an expression with that type.

(We could apply various constraints on expressions of type `List<void>`, `Map<int, void>` and similar types where `void` is a subterm, which is what we usually call 'voidness preservation', but we haven't introduced support for that concept yet. It is not needed for soundness because the protection of `void` expressions is purely a developer intention, there's nothing wrong with the object itself, even if it was obtained by evaluating an expression of type `void`, directly or indirectly.)

Moreover, the implementation of `toString` in that instance of `List<void>` has been compiled under the assumption that the type variable `E` has no constraints like those on `void`. So even though we actually have a `List<void>` here, at compile time, there is nothing stopping the `toString` implementation of that list from calling `toString` on each element.

So it's not a bug, and it doesn't even support the claim that "we really need to introduce voidness preservation now!", because voidness preservation wouldn't detect this situation. 

The actual reason why we fail to detect a problem is that we don't treat a type variable as if it were `void` even in the cases where it's bound is a top type (e.g., `Object?` or `dynamic`) such that it _could_ be `void`. We decided very early on that it would not be acceptable to treat every such type variable as if it were `void`, because it would introduce many errors in classes/functions taking type arguments with no bound, that is, it would probably be hugely breaking.

**Comment by nshahan** [comment]
Thanks for the explanation. All of these details make sense and in the end the reason I ran into this was because of some helpers in some test code so it wasn't leading to a problem in production. 

I just thought it was strange when I noticed a method that was mistakenly typed as returning void even though it was always returning a value. It was being called and the results were being used in test expectations... and it was all working fine. 

Since this isn't a bug, I'll close this issue.
